src/file_controller/mod.rs

```diff
@@ -85,50 +81,106 @@ impl Cache {
         vfs_err!(self.files.load_lines(path, line_start, line_end))
     }
 
-    pub fn get_highlighted(&self, path: &Path) -> Result<Vec<String>, String> {
+    pub fn get_highlighted(&self, path: &Path) -> Result<Highlighted, String> {
+        fn raw_lines(text: &str) -> Vec<String> {
+            let mut highlighted: Vec<String> = text.lines().map(|s| s.to_owned()).collect();
+            if text.ends_with('\n') {
+                highlighted.push(String::new());
+            }
+
+            highlighted
+        }
+
         vfs_err!(self.files.load_file(path))?;
         vfs_err!(
             self.files
                 .ensure_user_data(path, |_| Ok(VfsUserData::new()))
         )?;
         vfs_err!(self.files.with_user_data(path, |u| {
             let (text, u) = u?;
-            let text = match text {
-                Some(t) => t,
-                None => return Err(::vfs::Error::BadFileKind),
-            };
-            if u.highlighted_lines.is_empty() {
+
+            if u.highlighted.is_none() {
                 if let Some(ext) = path.extension() {
-                    if ext == "rs" {
-                        let highlighted = highlight::highlight(
-                            &self.analysis,
-                            &self.project_dir,
-                            path.to_str().unwrap().to_owned(),
-                            text.to_owned(),
-                        );
-
-                        let mut highlighted_lines = vec![];
-                        for line in highlighted.lines() {
-                            highlighted_lines.push(line.replace("<br>", "\n"));
+                    match ext {
```

Comment:
I think this would be clearer as an if/else rather than a match

Reply:
I think in the long run this would be best expressed as a `HashMap<&'static str, &'static SomeTrait>`